src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
+
+You can find an example for generating this in the [web-push node library](https://github.com/web-push-libs/web-push/). It looks something like this:
+
+{% highlight javascript %}
+function generateVAPIDKeys() {
+  var curve = crypto.createECDH('prime256v1');
+  curve.generateKeys();
+
+  return {
+    publicKey: curve.getPublicKey(),
+    privateKey: curve.getPrivateKey(),
+  };
+}
+{% endhighlight %}
+
+## Anatomy of a subscription object
+
+Earlier we said that a subscription object must be stringified and passed to the 
+server, but we didn't tell you what was in the subscription object. That's 
+because the client doesn't do anything with it. The server does.  
+  
+The subscription object looks like this:  
+
+{% highlight javascript %}
+{  
+  "endpoint": "https://example.com/push-service/send/dbDqU8xX10w:APA91b...",  
+  "keys": {  
+    "auth": "qLAYRzG9TnUwbprns6H2Ew==",  
+    "p256dh": 
+"BILXd-c1-zuEQYXH\\_tc3qmLq52cggfqqTr\\_ZclwqYl6A7-RX2J0NG3icsw..."  
+  }  
 }
 {% endhighlight %}
 
-What's in this?
+What's in this?  
+  
+**endpoint**—Contains two parts: the URL of the messaging service used by the
+subscribing browser followed by unique identifier for the user. This is called a
+subscription ID or  a registration ID. This tells your server how to identify
+you the messaging  server.
+  
+**keys**—Encryption keys used for encrypting data passed to the service worker 
+messages. It contains the following:
+
+* **auth**-A 16 byte authentication secret generated by the browser.
+* **p256dh**-A 65 byte containing the vapidPublicKey created on your app server.
+
+**Note:** In many of the relevant specs bytes are called octets. The term is 
+used because of legacy and communication systems where bytes may not always be 8 
+bits long.
+
+## Creating the message
+
+This is where things start to get a little crazy. In this section we're no 
+longer in your client app. We're on the app server where we are going to create 
+and send a message to the client. There's much to keep track of. 
+
+Before continuing, let's review what we have and where it came from.
+
+* **Subscription object** - his came from the client. It contains the endpoint 
+  of the messaging server, a copy of the public key (`applicationServerKey`), and an authentication 
+  secret generated by the client. From here on, we're going to stop talking 
+  about the subscription object and only refer to the **endpoint**, **public 
+  key**, and **authentication secret**.
+* **private key** - The private key corresponding to the public key. This is the private key you created in 'Generating the applicationServerKey'.
+
+We're going to look at message creation in three parts. First we'll create some 
+HTTP headers, then we'll create a payload for the message, and finally we'll 
+combine them and send them to the messaging server.
+
+### A note about code samples
```

Comment:
Not sure what I'm meant to take away from this, first pass made me think that this code would be very good / useful.